instrumentation/aws-sdk/aws-sdk-2.2/testing/src/main/java/io/opentelemetry/instrumentation/awssdk/v2_2/AbstractAws2ClientCoreTest.java

```diff
@@ -0,0 +1,403 @@
+            equalTo(stringKey("aws.table.name"), "sometable"),
+            equalTo(DB_SYSTEM, "dynamodb"),
+            equalTo(maybeStable(DB_OPERATION), "CreateTable"),
+            equalTo(
+                stringKey("aws.dynamodb.global_secondary_indexes"),
+                "[{\"IndexName\":\"globalIndex\",\"KeySchema\":[{\"AttributeName\":\"attribute\"}],\"ProvisionedThroughput\":{\"ReadCapacityUnits\":10,\"WriteCapacityUnits\":12}},{\"IndexName\":\"globalIndexSecondary\",\"KeySchema\":[{\"AttributeName\":\"attributeSecondary\"}],\"ProvisionedThroughput\":{\"ReadCapacityUnits\":7,\"WriteCapacityUnits\":8}}]"),
+            equalTo(stringKey("aws.dynamodb.provisioned_throughput.read_capacity_units"), "1"),
+            equalTo(stringKey("aws.dynamodb.provisioned_throughput.write_capacity_units"), "1"));
+  }
+
+  @SuppressWarnings("deprecation") // using deprecated semconv
+  private static void assertQueryRequest(SpanDataAssert span) {
+    span.hasName("DynamoDb.Query")
+        .hasKind(SpanKind.CLIENT)
+        .hasNoParent()
+        .hasAttributesSatisfyingExactly(
+            equalTo(SERVER_ADDRESS, "127.0.0.1"),
+            equalTo(SERVER_PORT, server.httpPort()),
+            equalTo(URL_FULL, server.httpUri() + "/"),
+            equalTo(HTTP_REQUEST_METHOD, "POST"),
+            equalTo(HTTP_RESPONSE_STATUS_CODE, 200),
+            equalTo(RPC_SYSTEM, "aws-api"),
+            equalTo(RPC_SERVICE, "DynamoDb"),
+            equalTo(RPC_METHOD, "Query"),
+            equalTo(stringKey("aws.agent"), "java-aws-sdk"),
+            equalTo(AWS_REQUEST_ID, "UNKNOWN"),
+            equalTo(stringKey("aws.table.name"), "sometable"),
+            equalTo(DB_SYSTEM, "dynamodb"),
+            equalTo(maybeStable(DB_OPERATION), "Query"),
+            equalTo(stringKey("aws.dynamodb.limit"), "10"),
+            equalTo(stringKey("aws.dynamodb.select"), "ALL_ATTRIBUTES"));
+  }
+
+  @SuppressWarnings("deprecation") // uses deprecated semconv
+  private static void assertDynamoDbRequest(SpanDataAssert span, String operation) {
+    span.hasName("DynamoDb." + operation)
+        .hasKind(SpanKind.CLIENT)
+        .hasNoParent()
+        .hasAttributesSatisfyingExactly(
+            equalTo(SERVER_ADDRESS, "127.0.0.1"),
+            equalTo(SERVER_PORT, server.httpPort()),
+            equalTo(URL_FULL, server.httpUri() + "/"),
+            equalTo(HTTP_REQUEST_METHOD, "POST"),
+            equalTo(HTTP_RESPONSE_STATUS_CODE, 200),
+            equalTo(RPC_SYSTEM, "aws-api"),
+            equalTo(RPC_SERVICE, "DynamoDb"),
+            equalTo(RPC_METHOD, operation),
+            equalTo(stringKey("aws.agent"), "java-aws-sdk"),
+            equalTo(AWS_REQUEST_ID, "UNKNOWN"),
+            equalTo(stringKey("aws.table.name"), "sometable"),
+            equalTo(DB_SYSTEM, "dynamodb"),
+            equalTo(maybeStable(DB_OPERATION), operation));
+  }
+
+  private static Stream<Arguments> provideArguments() {
```

Comment:
could use something like
```java
  @SuppressWarnings("unchecked")
  private static DynamoDbClient wrapClient(DynamoDbAsyncClient asyncClient) {
    return (DynamoDbClient)
        Proxy.newProxyInstance(
            DynamoDbClient.class.getClassLoader(),
            new Class<?>[] {DynamoDbClient.class},
            (proxy, method, args) -> {
              Method asyncMethod =
                  DynamoDbAsyncClient.class.getMethod(method.getName(), method.getParameterTypes());
              CompletableFuture<?> future =
                  (CompletableFuture<?>) asyncMethod.invoke(asyncClient, args);
              return future.get();
            });
  }
```
so that the parameters would not need to be duplicated between sync and async case. Async test could use `Object response = call.apply(wrapClient(client));` Note that the response passed to `validateOperationResponse` won't be a `Future` any more with this.
Could generalize this to
```java
  @SuppressWarnings("unchecked")
  protected static <T, U> T wrapClient(Class<T> syncClientClass, Class<U> asyncClientClass, U asyncClient) {
    return (T)
        Proxy.newProxyInstance(
            AbstractAws2ClientCoreTest.class.getClassLoader(),
            new Class<?>[] {syncClientClass},
            (proxy, method, args) -> {
              Method asyncMethod =
                  asyncClientClass.getMethod(method.getName(), method.getParameterTypes());
              CompletableFuture<?> future =
                  (CompletableFuture<?>) asyncMethod.invoke(asyncClient, args);
              return future.get();
            });
  }
```
from what I noticed only s3 `getObject` has different parameters for sync and async.

Reply:
you are a wizard. thank you for this suggestion